File: src/main/scala/org/raisercostin/jedi/VfsLocation.scala
package org.raisercostin.jedi

import scala.Iterable

import org.apache.commons.vfs2.FileObject
import org.apache.commons.vfs2.VFS
object VfsLocation{
  private val fsManager = VFS.getManager()
  def apply(url: String):VfsLocation = VfsLocation(fsManager.resolveFile(url))
}

case class VfsLocation(file:FileObject) extends NavigableInOutLocation { self =>
  override type Repr = self.type
  def raw = file.getName.getPath
  def fileFullPath: String = file.getName.getPath
  override def build(path:String): Repr = ???//new VfsLocation(file)
  def buildNew(x: FileObject): Repr = new VfsLocation(x)
  override def parent: Repr = buildNew(file.getParent)
  override def child(child: String): Repr = buildNew(file.getChild(child))
  override def childName(child:String):String = file.getChild(child).getName.getPath
  override def exists:Boolean = file.exists
  override def list: Iterable[Repr] = Option(existing).map { x =>
    Option(x.file.getChildren).map(_.toIterable).getOrElse(Iterable(x.file))
  }.getOrElse(Iterable()).map(buildNew)
  def isFile: Boolean = file.isFile
  def isFolder: Boolean = file.isFolder
  def mkdirOnParentIfNecessary:Repr = {
    file.getParent.createFolder
    this
  }
  def asInput: NavigableInOutLocation = self
  def append: Boolean = ???
  def withAppend: Repr = ???
  override def nameAndBefore: String = raw
  override def name:String = file.getName.getBaseName
  def external:VfsLocation = {
    var all = file.getURL.toExternalForm
    all = all.dropWhile { x => x!=':' }.drop(1)
    all = all.reverse.dropWhile{x=>x!='!'}.drop(1).reverse
    VfsLocation(all);
  }
  override def absolute:String = file.getName.getExtension
  override def toString = {
    "VfsLocation[url="+file.getURL+"]"
  }
  println(file)
  def withProtocol(protocol:String):VfsLocation = {
    val newUrl = protocol+":"+file.getURL.toString+"!/"
    println("new "+newUrl)
    VfsLocation(newUrl)
  }
  override def unsafeToInputStream = file.getContent.getInputStream
  override def unsafeToOutputStream = file.getContent.getOutputStream
}

File: src/main/kotlin/org/raisercostin/jedi/VfsLocation.kt
package org.raisercostin.jedi

import scala.Iterable

import org.apache.commons.vfs2.FileObject
import org.apache.commons.vfs2.VFS
object VfsLocation{
  private val fsManager = VFS.getManager()
  fun apply(url: String):VfsLocation = VfsLocation(fsManager.resolveFile(url))
}

data class VfsLocation(file:FileObject) : NavigableInOutLocation { self ->
  fun raw ()= file.getName.getPath
  fun fileFullPath: String = file.getName.getPath
  override fun build(path:String): self.type = ???//new VfsLocation(file)
  fun buildNew(x: FileObject): self.type = VfsLocation(x)
  override fun parent: self.type = buildNew(file.getParent)
  override fun child(child: String): self.type = buildNew(file.getChild(child))
  override fun childName(child:String):String = file.getChild(child).getName.getPath
  override fun exists:Boolean = file.exists
  override fun list: Iterable<self.type> = Option(existing).map { x ->
    Option(x.file.getChildren).map(_.toIterable).getOrElse(Iterable(x.file))
  }.getOrElse(Iterable()).map(buildNew) as Iterable<self.type>>
  fun isFile: Boolean = file.isFile
  fun isFolder: Boolean = file.isFolder
  fun mkdirIfNecessary: self.type {
    file.createFolder
    this
  }
  fun mkdirOnParentIfNecessary:self.type {
    file.getParent.createFolder
    this
  }
  fun asInput: NavigableInOutLocation = self
  override fun append: Boolean = ???
  fun ,Append: self.type = ???
  override fun size:Long = file.getContent.getSize
  override fun nameAndBefore: String = raw
  override fun name:String = file.getName.getBaseName
  fun external:VfsLocation {
    var all = file.getURL.toExternalForm
    all = all.dropWhile { x -> x!=':' }.drop(1)
    all = all.reverse.dropWhile{x->x!='!'}.drop(1).reverse
    VfsLocation(all);
  }
  override fun absolute:String = file.getName.getExtension
  override fun toString {
    "VfsLocation<url="+file.getURL+">"
  }
  println(file)
  fun ,Protocol(protocol:String):VfsLocation {
    val newUrl = protocol+":"+file.getURL.toString+"!/"
    println("new "+newUrl)
    VfsLocation(newUrl)
  }
  override fun unsafeToInputStream ()= file.getContent.getInputStream
  override fun unsafeToOutputStream ()= file.getContent.getOutputStream
}